users/vieting/experiments/switchboard/ctc/feat/network_helpers/specaug_configurable.py

```diff
@@ -106,13 +117,93 @@ def add_specaug_layer(
     # Update config with pre-generated parameters
     full_config["specaug_params"] = specaug_params
 
-    network[name] = {
-        "class": "eval",
-        "from": from_list,
-        "eval": f"self.network.get_config().typed_value('transform')(source(0, as_data=True), network=self.network, **{full_config})",
-    }
+    if not full_config["mask_based_on_variance"]:
+        network[name] = {
+            "class": "eval",
+            "from": from_list,
+            "eval": f"self.network.get_config().typed_value('transform')(source(0, as_data=True), network=self.network, **{full_config})",
+        }
+        return [name], [
+            sort_filters_by_center_freq,
+            get_frequency_response,
+            _mask,
+            random_mask,
+        ]
+    else:
+        network[name] = {
+            "class": "eval",
+            "from": from_list,
+            "eval": f"self.network.get_config().typed_value('transform_with_variance')(source(0, as_data=True), network=self.network, **{full_config})",
+        }
+        return [name], [
+            get_frequency_response,
+            calculate_frequency_response_variance,
+            _mask,
+            random_mask,
+            filter_based_masking ,
+            transform_with_variance,
+        ]
+
+
+def calculate_frequency_response_variance(filter_layer):
+    """
+    Calculate the variance of the frequency response for each feature.
+
+    :param tf.Tensor filter_layer: The filter layer tensor
+    :return: Variance of frequency response for each feature
+    :rtype: tf.Tensor
+    """
+    import tensorflow as tf
+
+    f_resp = get_frequency_response(filter_layer)
 
-    return [name], get_specaug_funcs()
+    # Calculate variance for each feature
+    variance = tf.math.reduce_variance(f_resp, axis=0)
+    return variance
+
+
+def filter_based_masking (x, batch_axis, axis, probability_distribution, max_number_masks_for_probability_specaug):
+    """
+    :param tf.Tensor x: (batch,time,feature)
+    :param int batch_axis:
+    :param int axis:
+    :param int max_number_masks_for_probability_specaug:
+    """
+    import tensorflow as tf
+    n_batch = tf.shape(x)[batch_axis]
+    ndim = x.get_shape().ndims
+    n_features = tf.shape(x)[axis]
+    
+    num_masks = tf.random.uniform(
+        shape=[n_batch],
+        minval=0,
+        maxval=max_number_masks_for_probability_specaug + 1,
+        dtype=tf.int32
+    )
+    max_masks = tf.reduce_max(num_masks)
```

Comment:
Why reduce max? You could have different numbers for each sequence in the batch. I think that's also the case for the baseline SpecAugment.